graph/patterns/common-query-functions.md

```diff
@@ -0,0 +1,119 @@
+# Exposing functions for common queries
+
+Microsoft Graph API Design Pattern
+
+*Exposing functions for common queries is providing functions with friendly names as an additional mechanism for clients to make certain kinds of requests that involve the use of OData query options.*
+
+## Problem
+
+Some APIs are very data-focused.
+These APIs are generally capable of providing robust `$fitler`ing and `$expand`ing functionality.
+However, the OData query options that need to be used by clients can become complicated and confusing.
+Because of this, API producers have a tendency to avoid supporting query options that require client developers to have OData experience; avoiding these query options results in APIs that are not fully featured and are not externally extensible, limiting what clients can actually accomplish via the APIs.
+
+## Solution
+
+The solution to this problem is to provide support for the query options and to additionally define OData functions that act as aliases for a specific set of query options.
+Doing this gives client developers a low barrier-of-entry to using the API (familiarizing themselves with the shape of the data), and it also gives those developers extensiblity in the future to grow beyond the basic cases to write clients that are specific to their own business needs.
+The functions can further provide discoverability by using names for concepts that customers are familiar with, letting them know that those concepts are supported.
+
+## When to use this pattern
+
+This pattern can be employed in almost any circumstance.
+It is also able to be used once an API as shipped and in any order.
+It is ok to ship a data-focused API with robust `$filter`ing support, and *later* ship functions that act in the same way as certain specific filters.
+It is likewise fine to ship a handful of functions and then later ship an API that is a collection of data with a more fully-featured set of filters.
```

Comment:
This makes it sound like it's fine to ship a special purpose function because, if in the future they decide they want something more flexible, they could add it. I'm afraid that would lead to workloads using this guidance to say "I have a bunch of functions today, and might add more in the future, cause the guidance says that I can always add $filter later."

I would think that we would want to encourage workloads to move to the more general solution and not rely on exposing everything through custom functions.

```suggestion
While it is possible to start with one or two custom functions, you should have a plan to support a more fully-featured set of query options as the utility of your API will be significantly limited in the interim.
```